tests/rest/client/test_retention.py

```diff
@@ -45,50 +45,66 @@ def make_homeserver(self, reactor, clock):
         }
 
         self.hs = self.setup_test_homeserver(config=config)
+
         return self.hs
 
     def prepare(self, reactor, clock, homeserver):
         self.user_id = self.register_user("user", "password")
         self.token = self.login("user", "password")
 
-    def test_retention_state_event(self):
-        """Tests that the server configuration can limit the values a user can set to the
-        room's retention policy.
+        self.store = self.hs.get_datastore()
+        self.serializer = self.hs.get_event_client_serializer()
+        self.clock = self.hs.get_clock()
+
+    def test_retention_event_purged_with_state_event(self):
+        """Tests that expired events are correctly purged when the room's retention policy
+        is defined by a state event.
         """
         room_id = self.helper.create_room_as(self.user_id, tok=self.token)
 
+        # Set the room's retention period to 2 days.
+        lifetime = one_day_ms * 2
         self.helper.send_state(
             room_id=room_id,
             event_type=EventTypes.Retention,
-            body={"max_lifetime": one_day_ms * 4},
+            body={"max_lifetime": lifetime},
             tok=self.token,
-            expect_code=400,
         )
 
+        self._test_retention_event_purged(room_id, one_day_ms * 1.5)
```

Comment:
I think I'm missing something with this test -- we're setting the lifetime to 2 days, but then only incrementing 1.5 days. But we expect it to be purged still? I don't think the allowed lifetime values come into play here since it is squarely in the middle. Why do we expect this to be purged?

Reply:
Actually, looks like my new test might be succeeding on accident. The way `_test_retention_event_purged` works is:

1. Send event 1
2. Increment time by `increment`
3. Send event 2
4. Increment time by `increment`
5. Check that event 1 has been purged
6. Check that event 2 has not been purged

So `increment` must be a value so that events aren't purged if they are `increment` old but should be purged if they are `increments * 2` old.

I'll correct my test and comment `_test_retention_event_purged` with this info.

Reply:
Ah! It increments that twice, that's what I was missing. 👍  Thanks for adding comments.